workspaces/src/error/mod.rs

```diff
@@ -0,0 +1,89 @@
+//! All errors that can occur within workspaces, including but not limited to
+//! the following: IO, RPC, Execution, Sandbox, DataConversion errors.
+
+mod impls;
+
+use std::borrow::Cow;
+
+/// A list specifying general categories of NEAR workspace error.
+#[derive(Clone, Debug, PartialEq, thiserror::Error)]
+#[non_exhaustive]
+pub enum ErrorKind {
+    /// An error occurred while performing an RPC request.
+    #[error("Rpc({0})")]
+    Rpc(#[from] RpcErrorCode),
+    /// An error occurred while processing a transaction.
+    #[error("Execution")]
+    Execution,
```

Comment:
Intuitively seems like someone might want to check that the execution error was a specific string/substring. We can always keep the API minimal and add later, though

Reply:
I would say it would be wrong to use `ErrorKind::Execution` to get details about execution. tx execution outcome is not a `Result<T, E>` -- even a failed transaction spends gas. The API should push the users who wish to inspect execution errors to use `CallExecutionDetails` directly. If the user need to see what the error was for debugging purposes, there's `fmt::Display` for that

Reply:
I suppose my comment could be targeted more specifically at the issue that `CallExecutionDetails` execution errors get automatically converted into the `workspaces::Result`, which is why I indicate this as a motivation. For example, this test passes:

```rust
#[test(tokio::test)]
async fn test_failure() -> anyhow::Result<()> {
    let worker = workspaces::sandbox().await?;
    let account = worker.dev_create_account().await?;

    let res = account.call(&worker, account.id(), "doesn't_exist").transact().await;
    assert!(matches!(res.unwrap_err().kind(), workspaces::error::ErrorKind::Execution));

    Ok(())
}
```

and it gives developers no way of inspecting the error message other than going through formatting the error into a string somehow. I generally think the API holistically needs to be more consistent and streamlined because there are a few inconsistencies that will make it harder for devs to use.

Reply:
Hm, that shoudn't automatically call `try_into_result`, that should be done by the user: 

```rust
let res = account.call(&worker, account.id(), "doesn't_exist").transact()
  .await? // this can fail if the sandbox is down
  .success()? // this will fail if the tx got to the network, but was faulty
;
```

`CallExecutionDetails` should be `#[must_use]`, and it shouldn't have data-erasign API like

```
fn try_into_result(self) -> Result<Self>
```

good catch!

Reply:
It's not only the question of error message: the user should be able to tell, eg, how much gas was spend by a failed transaction.

Reply:
https://github.com/near/workspaces-rs/pull/150 was gonna resolve the `#[must_use]` so I'll leave it for that after I sync it up with all the recent changes from this current PR